src,lib: make ^C print a JS stack trace

legendecas opened the pull request:
If terminating the process with ctrl-c / SIGINT, prints a JS stack trace leading up to the currently executing code.

The feature would be enabled under option `--trace-sigint`.

Conditions of no stacktrace on sigint:

1. has (an) active sigint listener(s);
2. main thread is idle (i.e. uv polling), a message instead of stacktrace would be printed.

Related: https://github.com/nodejs/node/issues/24937

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented on lib/internal/main/run_main_module.js:
```diff
@@ -3,11 +3,31 @@
+  // Keyboard interruption may not be captured by uncaughtException
+  // listener.
+  console.error(err);
+  process.exit(130);
```
Is this exit code correct for SIGINT across all platforms we support?

mscdex commented:
Maybe we should only print a stack trace if stdin is a TTY (especially if it's meant to only happen with pressing ctrl-c)?

addaleax commented on src/node_watchdog.cc:
```diff
@@ -90,6 +90,30 @@ void Watchdog::Timer(uv_timer_t* timer) {
+  auto callback = args[0].As<v8::Function>();
+  auto watchdog = new SigintWatchdog(isolate, callback);
+  self->Set(env->context(), env->owner_symbol(),
+            v8::External::New(isolate, watchdog));
```
Is this symbol property used somewhere? Where is the `new SigintWatchdog(...)` allocation released?

What you probably want to do is to create a subclass that inherits from `SigintWatchdog` and `HandleWrap` that deals with the additional features here, and has the `uv_async_t` as a member rather than a pointer to it. (You could make `HandleSigint()` virtual in that case and override it.)

(And, side note, I’d have a preference for not using `auto` everywhere because it makes the code easier to write but less readable/reviewable.)

bnoordhuis replied:
We don't have a strong policy on `auto` though, it's basically up to the PR author. As long as the RHS makes it clear enough what the type is, I don't have a problem with it.

addaleax commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
 
 SigintWatchdog::~SigintWatchdog() {
+  if (loop_activate != nullptr) {
+    uv_close(reinterpret_cast<uv_handle_t*>(loop_activate),
```
You’d want to use `env->CloseHandle(...)` rather than `uv_close()` here, so that the Environment can keep track of handles that are still closing, but that problem goes away if you switch to using `HandleWrap`.

addaleax commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+  CHECK_EQ(callback.IsEmpty(), false);
+  CHECK_EQ(callback->IsFunction(), true);
+  auto recv = v8::Object::New(isolate_);
+  callback->Call(callback->CreationContext(), recv, 0, {});
```
You don’t technically need a receiver here.

If you do take the above suggestion and switch to using `HandleWrap()`, the typical pattern would be to call a method on the object rather than passing one to the constructor, e.g.

```c++
MakeCallback(env->oninterrupt_string(), 0, nullptr);
```

and in JS

```js
const watchdog = new SigintWatchdog();
watchdog.oninterrupt = () => {
  ...
};
```

(i.e. in that case you don’t need to manage `ref` explicitly.)

addaleax commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+    isolate_->TerminateExecution();
+  } else {
+    uv_async_send(loop_activate);
+    isolate_->RequestInterrupt([](v8::Isolate* isolate, void* data) {
```
I don’t think it’s technically legal to do this (call JS from an interrupt handler), but we also already do it for the inspector? I’d at least check that the two don’t get in each others way, e.g. that you can set a breakpoint inside the JS handler or something…

bnoordhuis replied:
I believe it's allowed. The relevant code is here:

https://github.com/nodejs/node/blob/edc83a9203c2a8d0fd219aac95fd1615e001c484/deps/v8/src/execution/isolate.cc#L1429-L1445

Note the explicit switch to `VMState<EXTERNAL>`.

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -90,6 +90,30 @@ void Watchdog::Timer(uv_timer_t* timer) {
+  auto isolate = args.GetIsolate();
+  auto env = node::Environment::GetCurrent(args);
+  auto self = args.This();
+  auto callback = args[0].As<v8::Function>();
```
Please add a `CHECK(args[0]->IsFunction());` before this line. `args[0].As<v8::Function>()` is undefined behavior when `args[0]` isn't actually a function.

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+  v8::Isolate* isolate, Local<v8::Function> interrupt)
+    : isolate_(isolate) {
+  ref.Reset(isolate, interrupt);
+  auto env = node::Environment::GetCurrent(isolate);
```
Pass the environment explicitly instead of looking it up through the isolate.

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+    : isolate_(isolate) {
+  ref.Reset(isolate, interrupt);
+  auto env = node::Environment::GetCurrent(isolate);
+  loop_activate = new uv_async_t;
```
Prefer `new uv_async_t()` (note the parens), that creates a zeroed instance.

jasnell replied:
```suggestion
  loop_activate = new uv_async_t();
```

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+  loop_activate->data = this;
+  // Just activate the loop, doesn't do callback.
+  uv_async_init(env->event_loop(), loop_activate, [](uv_async_t* it){
+    SigintWatchdog* self = reinterpret_cast<SigintWatchdog*>(it->data);
```
`static_cast` is sufficient. Also on line 163.

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+  uv_async_init(env->event_loop(), loop_activate, [](uv_async_t* it){
+    SigintWatchdog* self = reinterpret_cast<SigintWatchdog*>(it->data);
+    self->HandleInterrupt();
+  });
```
Can you `CHECK_EQ(0, uv_async_init(...))` the return value? It's unlikely but not impossible for it to fail.

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
+    *received_signal_ = true;
+    isolate_->TerminateExecution();
+  } else {
+    uv_async_send(loop_activate);
```
Ditto, can you check that the return value is zero?

addaleax commented on src/node_watchdog.h:
```diff
@@ -55,24 +56,53 @@ class Watchdog {
-class SigintWatchdog {
+class SigintWatchdogBase {
+ public:
+  virtual bool HandleSigint() {
```
```suggestion
  virtual ~SigintWatchdogBase() {}
  virtual bool HandleSigint() = 0;
```

The virtual destructor is a best practice for base classes, and using `= 0` forces subclasses to override the method at compile time.

addaleax commented on src/node_watchdog.h:
```diff
@@ -55,24 +56,53 @@ class Watchdog {
+  SET_SELF_SIZE(JsSigintWatchdog)
+
+ private:
+  v8::Global<v8::Function> ref;
```
```suggestion
```

(I think.)

addaleax commented on src/node_watchdog.cc:
```diff
@@ -107,11 +115,92 @@ SigintWatchdog::~SigintWatchdog() {
+bool SigintWatchdog::HandleSigint() {
   *received_signal_ = true;
   isolate_->TerminateExecution();
+  return /** stop propagation */true;
```
If you want to be explicit here about the effects of the return value, you can use a two-valued `enum` instead of `bool`. (This is just fine too, of course.)

addaleax commented on lib/internal/main/run_main_module.js:
```diff
@@ -3,11 +3,36 @@
 
 const CJSModule = require('internal/modules/cjs/loader').Module;
 
+const watchdog = new SigintWatchdog();
```
This always remains alive until the end of the process, right? (Just making sure, because otherwise there’s a `.close()` call that should be added somewhere.)

Also, maybe setup code should be part of `prepareMainThreadExecution()`, so that it is not only in effect when running `node <scriptname>`?

legendecas replied:
Yes, it would last until the end of the process. I'm just thinking about attaching it to process instance to make this explicit and enable works like 27406.

addaleax replied:
I think that might open a whole bunch of issues, so maybe not in this PR?

legendecas replied:
So I'm just gonna attach it to the global process instance and leave it there, would this be good?

addaleax replied:
The thing about attaching it to the `process` object is that it effectively makes `SigintWatchdog` public API, and I’d be a bit reluctant to do that. I think the code here is fine as it is, I’m only really concerned about under which scenarios we should enable this feature

legendecas commented:
> `report:` is the subsystem for the node-report feature, i.e. `node --experimental-report` … this touches a related topic, but I think it might be confusing to name it that way

Yes, this work doesn't touch `report` specifically in codebase. Yet I don't find a more proper subsystem here https://github.com/nodejs/core-validate-commit/blob/master/lib/rules/subsystem.js . Since both `src` and `lib` are touched in this work, neither of those one might be appropriate IMO.

addaleax commented:
@legendecas You can also use `src,lib: ...`. :slightly_smiling_face: Honestly, I don’t have a better idea than that either.

addaleax commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -158,6 +165,31 @@ function setupSignalHandlers() {
+      // handling of SIGINT.
+      return;
+    }
+    if (process.stdout.isTTY) {
```
I think if we print using `console.error()` (which I agree with is correct, we should be checking stderr here

```suggestion
    if (process.stderr.isTTY) {
```

(Although we could maybe also move this check to the beginning of `setupStacktracePrinterOnSigint()` and skip the setup altogether in that scenario?)

legendecas commented:
I think this PR is out of WIP, ready to be reviewed thoroughly :)

addaleax commented on src/node_watchdog.h:
```diff
@@ -55,24 +61,50 @@ class Watchdog {
   bool* received_signal_;
 };
 
+class JsSigintWatchdog : public HandleWrap, public SigintWatchdogBase {
```
Just curious, was switching the bases necessary for the postmortem metadata test? That would probably point to a design problem with the postmortem metadata.

legendecas replied:
No, I just found existing code putting HandleWrap first so I would like keep it consistent since there is no strong opinion about it. 

The postmortem metadata test was affected by handles initiated during`pre_execution`. Actually if I got it right, case `DebugSymbolsTest, HandleWrapList` was expected to find the last item of the handle_wrap_queue for the newly initiated handle wrap instance.

addaleax replied:
Okay, thanks for clarifying that – the order of the classes should not make a difference :)

addaleax commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1980/

addaleax commented:
@legendecas Looks like there are still issues with the postmortem tests…

legendecas commented:
> @legendecas Looks like there are still issues with the postmortem tests…

I'll try to fix it :P

mscdex commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -158,6 +165,36 @@ function setupSignalHandlers() {
+function setupStacktracePrinterOnSigint() {
+  // Use tty_wrap directly to produce minimal require footprint.
+  const { isTTY } = internalBinding('tty_wrap');
+  if (!isTTY(/** stderr */2)) {
```
```suggestion
  if (!isTTY(process.stderr.fd)) {
```

addaleax replied:
This code should not have to instantiate process.stderr.

devsnek replied:
to clarify, instantiating stdio has side effects as it starts buffering.

mscdex commented on lib/internal/watchdog.js:
```diff
@@ -0,0 +1,59 @@
+     * been activated. Also make sigint removeListener been ran after signal
+     * wrap been stopped.
+     */
+    process.prependListener('newListener', this._onNewListener);
```
The event handler needs to be bound to the instance, otherwise `this` in the handler will point to `process`.

mscdex replied:
Alternatively, instead of using a class for all of this, just use local scope for all of the state since only one instance of the class is being created anyway. That way you don't have to worry about `this` values being correct or performance issues with classes.

addaleax replied:
They are bound to the instance.

mscdex replied:
@addaleax Can you explain? Currently `this` in `_onNewListener()` will point to `process` (instead of the `SigintWatchdog` instance) because event handlers are always called in the context of the emitter. That does not seem like the intended behavior here.

addaleax replied:
`_onNewListener()` is an arrow function, which means that `this` is automatically bound to the `SigintWatchdog` instance.

If you’d prefer to avoid this syntax for readability reasons, I think that’s fair – it’s a bit unusual, in part because it’s a new language feature, even if it works fine here, and I definitely had to read this more than once to make sure it’s right.

(I wouldn’t give performance considerations too much weight here because this is not a hot path, it only runs once per process at most.)

mscdex commented on lib/internal/watchdog.js:
```diff
@@ -0,0 +1,59 @@
+
+class SigintWatchdog extends JsSigintWatchdog {
+  _started = false;
+  _effective = false;
```
Properties like this are still very much unoptimized in V8, so it'd be better to set them in a `constructor()`.

legendecas replied:
Having benchmarked on initiating class fields with new class field syntax and constructors, I found that there is no significant performance difference on accessing (getting or writting) the initiated fields but does have not trivial difference on `new`ing instances (0.1ms vs 0.001ms). Yet considering SigintWatchdog doesn't get instantiated continuously, I'd wonder keeping the code easy to read might be more appropriate. What do you think?

legendecas commented:
@addaleax I'd guess stderr was redirected on jobs of ci.nodejs.org, thus SigintWatchdog was not enabled on it. I've modified the test case to simply find the last item in the handle_wrap_queue to address the differences on both conditions.

legendecas commented:
Looks like CI failed for unrelated cases.

node-test-commit-osx failed for newly added case `parallel/test-http-server-keepalive-req-gc`; 
node-test-commit-freebsd failed for `parallel/test-stdout-close-catch` and `parallel/test-worker-message-port-message-before-close`; 
node-test-commit-windows-fanned  failed for `parallel/test-child-process-fork-exec-path`.

addaleax commented:
@nodejs/collaborators This should be ready – any thoughts? Should this be semver-major? (I’d tend towards “yes”.)

mcollina commented:
I have a couple of questions. Is this going to print the stacktrace only for non-responsive scripts or for all of them?

addaleax commented:
> Is this going to print the stacktrace only for non-responsive scripts or for all of them?

All of them, as long as stderr refers to a TTY.

addaleax commented:
I would think it’s very helpful to have a stack trace available by default, and the usefulness of that outweighs the downside of printing more information in situations where it’s not necessary. An extra CLI flag might be nice for introducing this as semver-minor, but I’d really aim for enabling this as a sensible default.

That being said – @mcollina, would you feel more comfortable if this only affected non-idle applications that are currently running JS? I think the stack trace isn’t very helpful anyway when it’s waiting on the event loop.

legendecas commented:
> One of the feature I love the most about Node.js is to be able to press ctrl-c to terminate idle servers, and it would be annoying/scaring if I got an error displayed.

@mcollina  One other thing you mentioned that errors displayed on ctrl-c would be scaring, would it be more comfortable if we make the stacktrace displayed as:
```
KeyboardInterrupt: Script execution was interrupted by `SIGINT`
    at main (*/sigint_print_stack.js:*)
    at Object.<anonymous> (*/sigint_print_stack.js:*)
    at *
```

No aggressive error word.

mcollina commented:
> That being said – @mcollina, would you feel more comfortable if this only affected non-idle applications that are currently running JS? I think the stack trace isn’t very helpful anyway when it’s waiting on the event loop.

Yes, definitely. I'd be +1 on that change. Essentially if the user is interrupting a `while (true)` loop knowing the stack trace is valuable information.

---

@legendecas I think a stack trace is scaring by itself, and it synonym of "something that went wrong". Howewever, nothing actually went wrong if the process is actually idle.

---

On a side note, are we adding some API to implement similar behavior in a userland module? In other terms, could we add this information to the `'SIGINT'` emitted event?

legendecas commented:
> On a side note, are we adding some API to implement similar behavior in a userland module? In other terms, could we add this information to the 'SIGINT' emitted event?

In short, adding this information to process emitted 'SIGINT' event would be a behavior change.

SigintWatchdog implemented and used in this PR as exposed as internal module has a different behavior compared to `process#SIGINT` event. As SigintWatchdog would interrupt current JS execution and insert a new piece of code to determine what to do on signal -- it has a much eager intent on it's effect. Rather `process#SIGINT` is following current node handle's pattern, and would not interrupt current JS execution.

We _could_ expose this API, but not along with existing 'SIGINT' event IMHO. Also this would mean introducing vm interruption to public JS API.

legendecas commented:
Could I summarize that if the stack trace (with minor aggressive name `KeyboardInterrupt`) only be printed on active JS executing, but not on idling, would be more comfortable and acceptable?

@mcollina @Fishrock123

addaleax commented:
> Edit: Though I have a minor concerning that how could we tell if the process is idling or actively running js code?

I think our main options are either not using the `uv_async_t` and only relying on `RequestInterrupt[]` (if we receive something on an `uv_async_t`, we don't have a JS stack below that anyway) or using the idle-ness detection mechanism we use for the CPU profiler (grep around for `SetIdle` in the code).

> It might be confusing if sometime sigint would result to a interrupt stack trace but sometime nothing happen and exit silently.

Some shells give an indication of how a process aborted (e.g. interrupted), but alternatively we could also print something like ``KeyboardInterrupt: Script execution was interrupted by `SIGINT` `` without the stack trace.

legendecas commented:
> I think our main options are either not using the uv_async_t and only relying on RequestInterrupt[] (if we receive something on an uv_async_t, we don't have a JS stack below that anyway) or using the idle-ness detection mechanism we use for the CPU profiler (grep around for SetIdle in the code).

@addaleax The point I was trying to tell is while we sending `SIGINT`, it might be unstable on state idle or active. For example, a http server might be mostly idle but active if there is some incoming requests. There might be chances that we think it's active but it was just came to idle, vice versa.

The problem seems gone if we print a note in either conditions.

legendecas commented:
> I'm in agreement. This is useful information. The traces that Python dumps when I just want to kill a script... are not helpful.
> It's extra unhelpful if you're a novice, because you might think that you broke something!

@boneskull Only _proactive_ keyboard interaction ctrl-c(SIGINT) which is commonly used to interrupt program would result in a stacktrace printed. As a side note, program exited with SIGINT would have non-zero exit code, e.g. 130, which in some terminal may have _red_ indicators, both of which IMO already mean something went wrong.

mcollina commented:
> Could I summarize that if the stack trace (with minor aggressive name KeyboardInterrupt) only be printed on active JS executing, but not on idling, would be more comfortable and acceptable?

Yes.

benjamingr commented:
It is worth noting (as anyone in core running gyp a lot probably already knows) that other languages (and namely python) already do this by default.

I am for the changes suggested by anna/matteo regarding idle servers :]

boneskull commented:
@legendecas So it sounds like what I _don’t_ want to happen, yes? 

I mean, I don’t get a stack trace if I abort a long-running `find`, and that’s great, because I don’t need one. I really don’t want to show these more than absolutely necessary, and certainly not in red if we can help it

boneskull commented:
cc @bcoe

addaleax commented:
@legendecas So … from an implementation point of view, I think the easiest way to go about this is to add a boolean argument to `JsSigintWatchdog::HandleInterrupt()`, that is passed along to JS land and indicates whether the interrupt call comes from the `uv_async_t` or the `Isolate::RequestInterrupt()` callback, and which determines whether to print a stack trace or not.

I don’t think we can reliably tell the difference between a server that just happens to execute code currently or that is actually doing long-running CPU work. If you do want to implement something in that direction, I think the best way would be something that delays calling `Isolate::RequestInterrupt()` by a few milliseconds in order to give the `uv_async_t` a better chance to fire first.

joyeecheung commented:
If we only prints the stack trace when there is JavaScript executing, I can imagine a bunch of confused issues about why this feature "mysteriously doesn't work from time to time", as this is not something you would typically look at the docs to find out the details about if you are an experienced Node.js user.

legendecas commented:
@joyeecheung Yeah, it is a confusing behavior. 

As an attempt to make the behavior more straightforward, a message in both conditions would be printed.
```
Interrupt [KEYBOARD_INTERRUPT]: Script execution was interrupted by `SIGINT`
```

And stackstrace would only be printed if JS is actively executing.
```
Interrupt [KEYBOARD_INTERRUPT]: Script execution was interrupted by `SIGINT`
    at main (*/sigint_print_stack.js:*)
    at Object.<anonymous> (*/sigint_print_stack.js:*)
```

So there would be no silent exit on SIGINT (or manually listen on `process#SIGINT` to mute this feature).

addaleax commented:
@Fishrock123 @mcollina Can you take another look?

mcollina commented:
Can we simplify

`Interrupt [KEYBOARD_INTERRUPT]:`

with

`Keyboard Interrupt`?

mcollina commented:
I would prefer to have a few more @nodejs/tsc reviews before landing anyway.

I've tagged this semver-major as it's a pretty significant change. Also adding tsc-agenda, as it's likely something that we should discuss.

legendecas commented:
> Can we simplify
> 
> `Interrupt [KEYBOARD_INTERRUPT]:`
> 
> with
> 
> `Keyboard interrupt`?

Yes, I was just keeping consistent with current error & code style. Since there is no chance to capture the interrupt _error_ so far, maybe it would be fine with the simplified style?

mcollina commented:
> Yes, I was just keeping consistent with current error & code style. Since there is no chance to capture the interrupt error so far, maybe it would be fine with the simplified style?

+1

bnoordhuis commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -158,6 +165,45 @@ function setupSignalHandlers() {
+      Error.stackTraceLimit = 0;
+      const err = new KEYBOARD_INTERRUPT();
+      Error.stackTraceLimit = stackTraceLimit;
+      // util.inspect would wrap no stacktrace error with brackets.f
```
```suggestion
      // util.inspect would wrap no stacktrace error with brackets.
```

bnoordhuis commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -158,6 +165,45 @@ function setupSignalHandlers() {
+      const err = new KEYBOARD_INTERRUPT();
+      console.error(err);
+    }
+    /** re-raising sigint */
```
```suggestion
    // Re-raising SIGINT.
```

bnoordhuis commented on lib/internal/watchdog.js:
```diff
@@ -0,0 +1,67 @@
+     * Prepend sigint newListener to remove stop watchdog before signal wrap
+     * been activated. Also make sigint removeListener been ran after signal
+     * wrap been stopped.
+     */
```
Can you use `//` comments? Thanks.

legendecas replied:
Maybe we could enable eslint rule `multiline-comment-style` to enforce this style?

bnoordhuis commented on src/node_watchdog.cc:
```diff
@@ -107,11 +115,125 @@ SigintWatchdog::~SigintWatchdog() {
+          FIXED_ONE_BYTE_STRING(env->isolate(), #name); \
+    Local<v8::Number> sig_flag_val = v8::Number::New(env->isolate(), it); \
+    target->Set(env->context(), sig_flag_name, sig_flag_val) \
+      .Check(); \
```
Style nit: can you line up the backslashes? We usually (but not always) do that at the 79th column. `git grep '\\$' src/` for examples. :-)

legendecas replied:
I just noticed the rule `AlignEscapedNewlines` in file `/.clang-format`. It seems like we don't enforce the rules in `/.clang-format`?

addaleax replied:
The `clang-format` rules affect only `make format-cpp`, e.g. you can use `CLANG_FORMAT_START=master make format-cpp` for autoformatting.

bnoordhuis commented on src/node_watchdog.h:
```diff
@@ -35,6 +36,11 @@
 
+enum SignalPropagation {
+  ContinuePropagation,
+  StopPropagation,
```
Not exactly wrong but we'd usually either UPPER_CASE the enum values or write them as `kContinuePropagation`.

bnoordhuis commented on test/pseudo-tty/sigint_idle_print_note.js:
```diff
@@ -0,0 +1,28 @@
+} else {
+  /**
+   * Use inherited stdio child process to prevent test tools from determining
+   * the case as crashed from SIGINT
```
`//` comments in this and other tests, please.

boneskull commented:
@joyeecheung 

> If we only print the stack trace when there is JavaScript executing, I can imagine a bunch of confused issues about why this feature "mysteriously doesn't work from time to time", as this is not something you would typically look at the docs to find out the details about if you are an experienced Node.js user.

if we can't only show the stack when appropriate _and_ do it in a consistent manner, I'm :-1: on the feature.  I think it will confuse and/or annoy more people than it will help.

addaleax commented:
@joyeecheung @boneskull I’m confused – are you against the current state of the PR (showing a JS stack trace when JS is executing, only showing the name line when no JS is executing)/do you prefer something else?

joyeecheung commented on lib/internal/errors.js:
```diff
@@ -1224,3 +1239,6 @@ E('ERR_WORKER_UNSUPPORTED_EXTENSION',
 E('ERR_WORKER_UNSUPPORTED_OPERATION',
   '%s is not supported in workers', TypeError);
 E('ERR_ZLIB_INITIALIZATION_FAILED', 'Initialization failed', Error);
+E('KEYBOARD_INTERRUPT',
```
I see no need using the internal error system if this is not going to be captured anyway?

joyeecheung commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -158,6 +165,45 @@ function setupSignalHandlers() {
+      // util.inspect would wrap no stacktrace error with brackets.f
+      console.error(`${err.name} [${err.code}]: ${err.message}`);
+    } else {
+      const err = new KEYBOARD_INTERRUPT();
```
I see no need using the internal error system if this is not going to be captured anyway?

boneskull commented:
@addaleax 

1.  If we're to tailor the output based on some condition, I'm not sure "JavaScript is executing" is what we should be shooting for.  Often I will abort a script (read: _not_ a process listening on a port) partway through its execution.  I don't need, nor want, to see a stack trace there; it's not useful to me.  I believe _this very situation_ to be the most likely one.
1.  It sounds like @joyeecheung is suggesting that any such determination of when to show a stack trace or not would potentially be confusing.  
1. The original issue is something akin to "show stack trace when script is unresponsive".  I _would_ like a stack trace if my script seems "stuck" and I interrupt it--that seems like reasonable behavior--but is it impossible to determine in the general case? Do we make a best effort?  Does V8 allow this?
1. Listening for a `SIGUSR1` or other signal and generating a diagnostic report could be another way to get at this information; we already have this.
1. My main experience has been with Python exhibiting this behavior, and it's been almost entirely useless.
1. If conditionally showing the stack trace in any capacity turns out to be confusing (as I believe @joyeecheung is suggesting), then I'd just as soon maintain status quo or hide behind a flag.

joyeecheung commented:
@addaleax I still think the current behavior is confusing - not showing the stack trace can also be considered "buggy", unless there are some words explaining the behavior in the message.

boneskull commented:
I also imagine tool maintainers will get stack traces in bug reports which ultimately mean nothing other than "this is where the user aborted the process" or otherwise see the stack trace and believe it to be a bug.

joyeecheung commented:
I still prefer @mcollina 's previous suggestion:

> I think it will be good to have this behind a flag, as I see why it can be helpful to debug certain situations. In that case, I would not even check if it's a TTY.

Imagine the following cases when ^C is pressed:

1. The process is expected to run a lot of computations, the user just wants to terminate it early for some reason.
    1. The process is executing JS when ^C is pressed
        1. The user probably already knows what the stack trace looks like or they don't care, printing the message probably is not going to be valuable to them.
        1. The downside of printing the stack trace to stderr in this case is it might mislead people and they might think that something goes wrong.
    1. The process is not executing JS when ^C is pressed
        1. The user already knows that the script execution is interrupted so seeing the message doesn't help (though it might be helpful to beginners if message explains that the process terminates by default if you don't attach a handler) 
         1. The downside of having inconsistent behavior (not printing or printing something different) is that it could be considered buggy.
1. The process is not expected to be stuck in script execution
    1. The process is executing JS when ^C is pressed
         1. In this case printing the stack trace does help the user figure out what's going on
    1. The process is not executing JS when ^C is pressed
         1. Inconsistent behavior (not printing or printing something different) could be considered buggy.

I think 2-i-a is rare compared to other cases, where the current behavior is either not helpful or confusing. Also notice that when there are a lot of child processes being spawned, printing a lot of messages when one simply presses ^C could be quite overwhelming.

joyeecheung commented:
BTW I noticed this when trying out the PR:

```
node benchmark/compare.js --new out/Release/node --old out/Release/node cluster > out.csv
[00:00:03|%   0| 0/1 files |  0/60 runs | 1/4 configs]: cluster/echo.js ^C
tty.js:66
Object.setPrototypeOf(ReadStream.prototype, net.Socket.prototype);
                                                       ^

TypeError: Cannot read property 'prototype' of undefined
    at tty.js:66:56
    at NativeModule.compile (internal/bootstrap/loaders.js:300:5)
    at nativeModuleRequire (internal/bootstrap/loaders.js:189:14)
    at createWritableStdioStream (internal/process/stdio.js:154:17)
    at process.getStderr [as stderr] (internal/process/stdio.js:40:14)
    at Object.get (internal/console/constructor.js:158:40)
    at Object.Console.<computed> (internal/console/constructor.js:276:46)
    at Object.warn (internal/console/constructor.js:287:61)
    at interrupt (internal/bootstrap/pre_execution.js:197:15)
    at SigintWatchdog.hidden [as oninterrupt] (internal/errors.js:256:14)
Interrupt [KEYBOARD_INTERRUPT]: Script execution was interrupted by `SIGINT`
```

This could be related to https://bugs.chromium.org/p/v8/issues/detail?id=9068 ?

boneskull commented:
> Also notice that when there are a lot of child processes being spawned, printing a lot of messages when one simply presses ^C could be quite overwhelming.

:+1: and :+1: for a flag

addaleax commented:
> If we're to tailor the output based on some condition, I'm not sure "JavaScript is executing" is what we should be shooting for. 

@boneskull It’s a distinction that makes sense from a technical point of view – if JS is executing, there is a stack trace available, and otherwise, there isn’t (or at least no meaningful one).

> Often I will abort a script (read: _not_ a process listening on a port) partway through its execution.  I don't need, nor want, to see a stack trace there; it's not useful to me.  I believe _this very situation_ to be the most likely one.

I think you’re right about this being the a very common use case when interrupting Node.js.

> The original issue is something akin to "show stack trace when script is unresponsive".  I _would_ like a stack trace if my script seems "stuck" and I interrupt it--that seems like reasonable behavior--but is it impossible to determine in the general case? Do we make a best effort?  Does V8 allow this?

That seems quite directly like the halting problem, so no, that can’t be determined reliably. And I think any best effort approach would require delaying the process exit a bit – e.g. to sample stack traces and see if they are repetitive, or to give the program a chance to return to the event loop.

> Listening for a `SIGUSR1` or other signal and generating a diagnostic report could be another way to get at this information; we already have this.

That’s true, but to me the point of this PR is to make the information more easily accessible. (You also already had the option of attaching the inspector while the process is running).

> My main experience has been with Python exhibiting this behavior, and it's been almost entirely useless.

Fair point, I would agree about it usually not being useful.

> If conditionally showing the stack trace in any capacity turns out to be confusing (as I believe @joyeecheung is suggesting), then I'd just as soon maintain status quo or hide behind a flag.

Honestly, I don’t see a lot of extra value of this feature when put behind a flag. As you noted, the information is already available through other venues, so adding another way to get it that requires some extra setup seems redundant.

> @addaleax I still think the current behavior is confusing - not showing the stack trace can also be considered "buggy", unless there are some words explaining the behavior in the message.

@joyeecheung This feels like an issue that could be circumvented by being a bit more explicit? i.e. explicitly stating in the output that there was no JS running at that point, or using a stack trace that indicates that the event loop is “on top of the stack”?

> I also imagine tool maintainers will get stack traces in bug reports which ultimately mean nothing other than "this is where the user aborted the process" or otherwise see the stack trace and believe it to be a bug.

@boneskull Tbh, I’m not all that concerned about it. I would expect users to understand that the stack trace is a reaction to the action they just took (interrupting the process).

> Also note that when there are a lot of child processes being spawned, printing a lot of messages when one simply presses ^C could be quite overwhelming, so I think it's better to make this behavior opt-in.

@joyeecheung I think that’s also a very fair point. Spawning child processes with Node.js is more common than with Python.

> BTW I noticed this when trying out the PR:
> 
> […]
> 
> This could be related to https://bugs.chromium.org/p/v8/issues/detail?id=9068 ?

Thanks for digging up that issue! I think that resolves https://github.com/nodejs/node/pull/29207#discussion_r315652471 – it sounds like the V8 team generally is on board with continuing to allow JS execution in the interrupt callback, which makes sense to me.

> We are expecting the state of execution in the interrupt callback, it should help if console is loaded eagerly when this feature is on, but note that loading console and compiling all its dependency slows down startup significantly so the best option is still making this opt-in iMO.

I think those are two separate issues. We should write this feature in a way that doesn’t require loading console early on, not modify the functionality of it.

---

The two of you make some good points arguing against this feature. You don’t have to agree with my wording, but I’d like to get more of an intuition on how people feel about this, so I’ve started an (as always very unscientific) Twitter poll to get more input: https://twitter.com/addaleax/status/1169986561775157248

addaleax commented:
It seems like a lot of people would like to have this optional behind a flag, so I’d guess that’s the way to go here? I’m still worried about that reducing the value added by this feature, but if that’s what people feel most comfortable with, I’m not going to stand in the way of it :)

Fishrock123 commented:
I am not sure it is worth implementing behind a flag.

 I still think we should do this for JS execution. I believe the potential ease benefit outweighs the slight negatives that I understand to exist, mostly in the form of minor annoyance?

bnoordhuis commented:
An observation: I've heard pythonistas complain about this particular aspect of python exactly never. I'm confident that in practice the benefits will far outweigh the downsides.

mhdawson commented:
One option I've not heard mentioned yet (or maybe I missed it in the long thread) is on by default but a flag to disable.  If you really don't like it you could use NODE_OPTIONS in your shell config files to disable it.

legendecas commented:
> One option I've not heard mentioned yet (or maybe I missed it in the long thread) is on by default but a flag to disable. If you really don't like it you could use NODE_OPTIONS in your shell config files to disable it.

We could already disable the feature of the PR by manually listening on SIGINT, still I think an additional cli option or environ to disable it would be nice. +1 on that.

benjamingr commented:
What about making it opt-in with a flag initially and opt-out at a following major after we ask people to try it and provide feedback?

mhdawson commented:
@benjamingr I think opt-in, followed by opt-out could be a good strategy for getting initial feedback, Only concern is that it would potentially be 6 months before the major where it could be switched (14.x most likely?)

Fishrock123 commented:
If it must be opt-in, an ENV var would probably be more _usable_.

jasnell commented:
Giving it more thought, I would only be +1 if it is opt-in using a command-line flag, e.g. `--trace-sigterm` perhaps. I would also suggest that it should be experimental while potential issues are identified and resolved (@joyeecheung also suggested this possibility in the TSC meeting). We can then decide later to switch the default and make the command-line flag a non-op.

I am -1 on this being enabled by default at the current time.

mcollina commented:
Hey @nodejs/tsc, would you prefer to always print a stacktrace or a message:

* always 👍 
* behind an opt-in flag 🎉 
* always, with an opt-out flag 🚀 
* never 👎

joyeecheung commented:
I am +1 if this is experimental with a flag, either opt-in or opt-out, then we decide what the default behavior should settle on. That separates the discussion of the default behavior from the review of the implementation - mind that we still need to resolve the interrupt issue in https://github.com/nodejs/node/pull/29207#issuecomment-527667993

sam-github commented:
@joyeecheung is of the opinion that the current implementation isn't very robust, that is a big concern.

ruby and python both do this, so it isn't unreasonable to consider, but on the other hand, I ctrl-C things when I realize its not what I want to do, and I don't really want stack traces. Its not clear that ctrl-C is useful for debugging, services that are misbehaving usually are in the background, and get `kill`ed, which sends SIGTERM (by default). But despite not sounding particularly useful, it also doesn't seem particularly harmful.

Hiding the option as opt-in sounds not useful to me. It allows us to not decide for or against, at the cost of kicking it into the "exists, but no one will ever us it" category, particularly if its a CLI or NODE_ENV option, because those are usually used to configure background processes (and in those cases, hooking SIGTERM makes more sense). Its hard to see CLI users changing their env to opt-in so that ctrl-Cing node CLIs give tracebacks.

If we want feedback on whether its liked, landing it as semver-major in master, so its there for everyone using master or nightlies is reasonable, IMO. That way we can see if it annoys anyone, and if so we can remove it before the next current is released from master.

So, +/- 0 on doing this for SIGINT, +1 for doing it for SIGTERM (which also gives a chance to try it out to see if its useful), -0 for hiding it behind an opt-in flag.

benjamingr commented:
WhileI voted "opt out" - I asked this question in the Node.js Israel users group and users seemed to not like the opt-out behavior. This might be attributed to my poor polling skills, them not understanding the issue, the polling platform or other factors. Funnily enough the people who voted "write to stderr" are mostly Python developers:

<img width="388" alt="Screen Shot 2019-09-11 at 23 11 03" src="https://user-images.githubusercontent.com/1315533/64735882-bc1d0e00-d4e9-11e9-8e7a-0d0e50150af2.png">

When I asked the users about whether or not they use other opt-in flags like --trace-sync-io and I did not get the impression that they did.

targos commented:
I have the impression that most users don't know about our debugging CLI flags (trace-sync-io, trace-warnings, etc.)

mhdawson commented:
@nodes/tooling we'd like your input on this.

addaleax commented:
Fwiw, here are the results from my unscientific Twitter poll:
![twitterpoll](https://user-images.githubusercontent.com/899444/65183335-8e911100-da63-11e9-978d-4121af6a2a61.png)

MylesBorins commented:
While my personal bias is to put it behind a flag, I think the arguments for having it on by default are very reasonable.

As discussed in the TSC meeting here is a potential action plan

* gather feedback from @nodejs/tooling 
* Land in master on by default with a flag to opt out
* Aim to land in Node.js 13
* If there is community push back at any of the above phases be willing and open to switch the defaults of the flag

Thoughts?

jasnell commented:
I still believe it should be an opt-in and the feedback in this thread would give the appearance that is the majority opinion. I'd be all for changing the default from opt-in to opt-out later but there's strong potential in breaking existing stuff that I'd prefer to avoid.

boneskull commented:
opt-in _please_

benjamingr commented:
@addaleax I think we have discovered indeed that people are more likely to pick the first option in a poll 😅 

@boneskull 
> opt-in please

Please elaborate on why?

boneskull commented:
a good point made in today's tooling meeting was that as tooling authors, our users do not necessarily care what language the tool is implemented in.  showing JS stack guts to them would not be friendly or helpful when _they_ choose to interrupt the process.  stack trace on error, sure, that'd be helpful for them to paste into a bug report--but not from a sigint.

@coreyfarrell noted that because of Python's behavior, he's had to _write code to suppress the stack traces_.

boneskull commented:
to be clear, when I say "opt-in please", I mean that if we do this at all, please don't make it on by default.  I'd also be fine with just _not_ adding it, because it's unlikely I'll ever want to use this.

benjamingr commented:
@boneskull alternatively opt out could be made easier

yorkie commented:
**Opt-out** is more smooth for existing Node.js developers, but **opt-in** is obviously better to share this function. Therefore, could we set a certain condition, when it is satisfied, then switch from **opt-out** to **opt-in**?

cclauss commented:
> he's had to write code to suppress the stack traces.

I am struggling with this a bit...  This effort was about either:
1. showing _more_ information that the current implementation or
2. showing _the same_ information as the current implementation

This effort should _not_ be used to show _less_ information than the current implementation.

boneskull commented:
@cclauss hmm not sure we’re talking about the same thing. the example above was regarding python, where this behavior is (apparently difficult) to opt-out of. he rightly wanted to suppress these stack traces in a command-line app he was writing (in python). 

if this is implemented in node as opt-out, then many CLI tools will want to add boilerplate to disable it (regardless of how “easy” it is to opt-out of). 

making it opt-out will create a lot of minor headaches for a lot of tooling authors (and others!) that don’t want this behavior. 

this is the general feeling I got from our discussion in this weeks @nodejs/tooling meeting; this must be opt-in (or don’t add it).

coreyfarrell commented:
> the example above was regarding python, where this behavior is (apparently difficult) to opt-out of. he rightly wanted to suppress these stack traces in a command-line app he was writing (in python).

Sorry if I gave the wrong impression, less difficult more of a simple nuisance.  In python the way I know to suppress it involves wrapping the application execution in a `try` block with `except` catching the termination exceptions and just exiting.  It's not much code but I end up putting it in every script that spends any time executing.  I understand the ability to opt-in to this feature might be useful to some developers but I would never enable it.

The argument that it can be made easy to opt-out is countered by the argument that it can be made easy to opt-in.  Being opt-out would create a difficult conflict for any test runners that don't want this.  Generally when I press Ctrl-C on a test it's because a test failed.  I want that failure to stay in view, I do not want a backtrace to fill the console / scroll the test failure out of view.  So does @boneskull make mocha opt-out of this SIGINT handler?  What if someone does want the backtrace from a mocha run - how do they opt-in if mocha has already opted-out?

mhdawson commented:
Discussed again in TSC meeting today, consensus from the 10 TSC members present is opt-in. That seems to align with previous feedback as well as feedback from the tooling team.  Believe TSC feedback is that is should be opt-in.

mhdawson commented:
Removing from tsc-agenda as we believe we've answered the question posed.

legendecas commented:
Updated with an option `--experimental-sigint-stacktrace` to opt-in the feature, while removing the guard on if stderr is a TTY (since its an opt-in feature already). Also, fixed styles @bnoordhuis mentioned above.

@mcollina @Fishrock123 PTAL :)

legendecas commented:
With the feature changed to be opt-in, shall the PR be labeled as semver-minor?

addaleax commented:
@legendecas Any particular reason for going with an `--experimental-...` flag name? We usually only do that when there’s a clear path for moving something out of experimental status, I think. I’d personally prefer `--sigint-stacktrace` over it.

boneskull commented:
@coreyfarrell I should have qualified “difficult”; what I meant was to opt-out in python, boilerplate code is needed instead of a flag

legendecas commented:
> @legendecas Any particular reason for going with an `--experimental-...` flag name? We usually only do that when there’s a clear path for moving something out of experimental status, I think. I’d personally prefer `--sigint-stacktrace` over it.

It's true that the feature has no clear path to be opt-out, updated with removing the `experimental` prefix.

yorkie commented on doc/api/cli.md:
```diff
@@ -620,6 +620,13 @@ Enables report to be generated on un-caught exceptions, if
 `--experimental-report` is enabled. Useful when inspecting JavaScript stack in
 conjunction with native stack and other runtime environment data.
 
+### `--sigint-stacktrace`
```
A little naming suggestion: `--trace-sigint` to keep conventions at https://github.com/nodejs/node/blob/b43895cc3c8f05ea7dabc8dd1f222b73da107d4f/doc/api/cli.md#--trace-deprecation.

legendecas replied:
Sounds good to me.

legendecas commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -168,6 +169,53 @@ function setupSignalHandlers() {
+      // handling of SIGINT.
+      return;
+    }
+    const console = require('console');
```
I'll update the implementation soon. Thank you for the suggestion :)

cclauss commented:
So far this PR looks like code-only.  Is there documentation that helps Node.js users understand how this change effects them and how to turn the feature on/off, etc?

legendecas commented:
> @cclauss So far this PR looks like code-only. Is there documentation that helps Node.js users understand how this change effects them and how to turn the feature on/off, etc?

There is a statement on the opt-in option `--trace-sigint` in `doc/api/cli.md`. But not very detailed effects like how it would exactly work statements like other options.

antsmartian commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -168,6 +169,53 @@ function setupSignalHandlers() {
+    });
+    if (flag === SignalFlags.FromIdle) {
+      const stackTraceLimit = Error.stackTraceLimit;
+      Error.stackTraceLimit = 0;
```
why are we setting this value to `0` here? Any particular reason?

legendecas replied:
The stacktrace printed for SIGINT triggered during idle (i.e. uv-polling) isn't much related to the actual calling stack. So just disabling it.

antsmartian commented on lib/internal/bootstrap/pre_execution.js:
```diff
@@ -168,6 +169,53 @@ function setupSignalHandlers() {
+    process.kill(process.pid, 'SIGINT');
+  };
+  watchdog.start();
+  return watchdog;
```
nit: I believe we are not using the return value anywhere in the code. We can get rid of this line.

antsmartian commented:
I was trying to run this in debug build (for setting up lldb and debugging whats happening under the hood), I faced the following issue. 

Running from debug build and killing the process with `SIGINT`, is actually printing this exception:

```
#
# Fatal error in ../deps/v8/src/execution/frames.h, line 149
# Debug check failed: static_cast<uintptr_t>(type) < Type::NUMBER_OF_TYPES (2189457190 vs. 23).
#
#
#
#FailureMessage Object: 0x7ffeefbfc6a0Illegal instruction: 4
```

Not sure what's wrong here, but I couldn't able to get the gist of it. Something I'm doing wrong here, causing the issue?

legendecas commented:
@antsmartian do you have a minimal setup to reproduce your problem? I can't reproduce it with `node --trace-sigint -e 'while (true) {}'` and `node --trace-sigint -e 'setTimeout(() => {}, 2000)'` for multiple times (around 100 times).

Also, I noticed that the error comes from debug checks on frames which might be related to problems stated here https://github.com/nodejs/node/pull/29207#discussion_r329356885 . Please point out if I got wrong.

antsmartian commented:
@legendecas I was trying the code from: https://github.com/nodejs/node/issues/27406. Yeah, as mentioned in my comment -- It was happening once in a while. The same code works as expected (i.e printing the stacktrace on sigint with the flag on) but fails like the above one (when I tried). I certainly doesn't know when frame error occur's, but anyone with that knowledge can help us here.

addaleax commented:
@antsmartian If you could get a stack trace out of a core dump, that might be very helpful (`gdb node_g core` assuming the core dump file is named `core` + running `bt` inside gdb)

psmarshall commented:
Unfortunately I know this DCHECK well... it generally means we are trying to walk the JS portion of the stack and run into random garbage that we can't interpret. That shouldn't happen as we try to avoid the very specific cases where we can't walk the stack. A stack trace would be very helpful as Anna suggested.

antsmartian commented:
@addaleax @psmarshall I tried again the same code, but no luck. The program was running successfully and sigint isn't causing the crash as before I posted here. If I get the same exception again, will share the stacktrace here.

legendecas commented:
I've updated the PR to not call into JS on interruption. Though this would cause the stacktrace to be not able to be decorated if I got it right.

fhinkel commented:
@Fishrock123 @mcollina @joyeecheung Could you have a look if your comments got addressed please. Thanks

Trott commented:
@legendecas Can you rebase?

@Fishrock123 @mcollina @joyeecheung I *think* your change requests have been addressed. Can you confirm?

legendecas commented:
Rebased. Would like to hear feedback from @Fishrock123 @mcollina @joyeecheung 😄

joyeecheung commented on src/node_watchdog.h:
```diff
@@ -54,24 +61,54 @@ class Watchdog {
+
+  SignalPropagation HandleSigint() override;
+
+  SET_NO_MEMORY_INFO()
```
Can you implement the `MemoryInfo` method of this class since it has `uv_async_t`? You can check out `src/memory_tracker.h` on how to use it.

(on a side note, we probably want to write more stuff in the variant of `TrackField` that takes a `uv_async_t` so that the `UV_HANDLE_FIELDS` show up in heap snapshots for debugging)

legendecas replied:
I'm confused that `SET_SELF_SIZE` might already include the inlined uv_async_t handle, is it still required to track the field? IIUC the additional space shall be tracked manually if we're referencing it with a pointer, like `uv_async_t*`.

joyeecheung replied:
When you track a non-pointer field its size would be subtracted from the size of the parent node and a new node will be created and displayed in the snapshot. Therefore you can see its value in the heap snapshot (this becomes more useful when the handle is deeply tracked, then you can discover other handles that points to similar handles or data or group them)

legendecas replied:
Updated, please take a look again :D.

joyeecheung commented on src/node_watchdog.h:
```diff
@@ -54,24 +61,54 @@ class Watchdog {
+  SET_SELF_SIZE(TraceSigintWatchdog)
+
+ private:
+  enum SignalFlags { None, FromIdle, FromInterrupt };
```
Since you already use the namespace everywhere..

```suggestion
  enum class SignalFlags { None, FromIdle, FromInterrupt };
```

legendecas commented:
All CI is green now. @Fishrock123 @mcollina may I have your review on the PR again?

addaleax commented on src/node_watchdog.h:
```diff
@@ -24,16 +24,24 @@
-#include "node_mutex.h"
 #include <vector>
+#include "handle_wrap.h"
+#include "memory_tracker-inl.h"
```
We generally shouldn’t be including `-inl.h` files from `.h` files – Can you move the `MemoryInfo()` method into the `.cc`file? That should remove the need for it.

legendecas replied:
But I noticed that we usually include the "memory_tracker-inl.h" in most of the sources. 

I would like to open another PR for it, and following the conventions in this PR.

Trott commented:
I think dismissing the remaining objection under the "unresponsive" clause at https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#consensus-seeking is sensible. There have been 5 @-mention pings about it over the last 2+ months and no response.

Fishrock123 commented:
I will attempt to take a look but I am in transit today. If I don’t see it by mid Monday just go ahead.

Fishrock123 commented:
I don't have ay substantive opinions.

legendecas commented:
Since there are no objections anymore and multiple approves, I'd tag this as author-ready.

legendecas commented:
Rebased to resolve conflicts.

legendecas commented:
Landed with 7b7e7bd185aa606cf2845747dd3844a9b12ec9ab 🎉

ruyadorno commented:
/cc @nodejs/tooling this landed 🎉 yay! we might want to talk about it in the next tooling meeting

orgads commented:
Is it possible to extend this also for `sigterm` (with a separate flag)?
